AnnService/inc/Core/Common/cuda/Distance.hxx

```diff
@@ -46,16 +50,108 @@ template<> __forceinline__ __host__ __device__ float INFTY<float>() {return FLT_
 //template<> __forceinline__ __host__ __device__ __half INFTY<__half>() {return FLT_MAX;}
 template<> __forceinline__ __host__ __device__ uint8_t INFTY<uint8_t>() {return 255;}
 
+template<typename T, typename SUMTYPE, int Dim>
+__forceinline__ __device__ SUMTYPE cosine(T* a, T* b) {
+    SUMTYPE total[2]={0,0};
+    for(int i=0; i<Dim; i+=2) {
+      total[0] += ((SUMTYPE)(a[i] * b[i]));
+      total[1] += ((SUMTYPE)(a[i+1] * b[i+1]));
+    }
+    return 1-(total[0]+total[1]);
```

Comment:
cosine may need to use GetBase instead of 1

Reply:
Added GetBase function on GPU and use it here.